flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java

```diff
@@ -114,7 +135,12 @@ public void addSplitsBack(List<PulsarPartitionSplit> splits, int subtaskId) {
 
         // If the failed subtask has already restarted, we need to assign pending splits to it.
         if (context.registeredReaders().containsKey(subtaskId)) {
-            assignPendingPartitionSplits(singletonList(subtaskId));
+            LOG.debug(
+                    "Reader {} has been restarted after crashing, we will put splits back to it.",
+                    subtaskId);
+            // Reassign for all readers in case of adding splits after scale up/down.
+            List<Integer> readers = new ArrayList<>(context.registeredReaders().keySet());
+            assignPendingPartitionSplits(readers);
```

Comment:
Will addSplitsBack be called when scale up/down ? Can you elaborate when will this be called during scale up/down ? Thanks ~

Reply:
This is a good question. Flink will restart the pipeline every time we scale up/down.

`addSplitsback` is called only when we enable [Restart Pipelined Region Failover Strategy](https://nightlies.apache.org/flink/flink-docs-master/docs/ops/state/task_failure_recovery/#restart-pipelined-region-failover-strategy) and some readers failed. We will put these splits back to Enumerator and reassign them after the failed readers started.

Reply:
"Reassign for all readers in case of adding splits after scale up/down.", this sentences occurs to me that the method will be involved in the scale up/down process. What does this sentence implies ?

Reply:
Due to [auto-scaling](https://flink.apache.org/2021/05/06/reactive-mode.html#getting-started) the parallelism may change over time. This may led to some splits are [reassigned to other reader instance](https://github.com/streamnative/flink/blob/8088a0996ed3d7ac956ee4c10f7cbbc0cd410e0c/flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NonSharedSplitAssigner.java#L110). So we shall try to assign the pending splits to all the available reader instance.